### PR TITLE
Table SAS and SharedKey permissions

### LINKED ISSUE #495: [Table] Phase 1 Feature - Service SAS Authentication
Read through:

- https://docs.microsoft.com/en-us/rest/api/storageservices/create-service-sas
- [Azurite V3 Table wiki](https://github.com/Azure/Azurite/wiki/Azurite-V3-Table)

Implement following parts:

- [ ] Add AuthenticationMiddlewareFactory.ts into src/table/middlewares. Refer to [Azurite Blob AuthenticationMiddlewareFactory](https://github.com/azure/azurite/blob/2395dec0eb8f235922ba34633bd81e3b33704caf/src/blob/middlewares/AuthenticationMiddlewareFactory.ts)
- [ ] Add IAuthenticator.ts into src/table/authentication. Refer to [Azurite Blob IAuthenticator.ts](https://github.com/azure/azurite/blob/3a0c73bb50ea7408ad5bfa79df542d6b8f879c36/src/blob/authentication/IAuthenticator.ts)
- [ ] Add IAuthenticationContext.ts into src/table/authentication. Refer to [Azurite Blob IAuthenticationContext.ts](https://github.com/azure/azurite/blob/3a0c73bb50ea7408ad5bfa79df542d6b8f879c36/src/blob/authentication/IAuthenticationContext.ts)
- [ ] Add and implement TableSASAuthenticator.ts into src/table/authentication. Refer to [Azurite Blob BlobSASAuthenticator.ts](BlobSASAuthenticator.ts).
- [ ] Add test cases. Refer to [Azurite Blob SAS tests](https://github.com/azure/azurite/blob/efe4e9d123304715cb84499653e008cc425d15fd/tests/blob/sas.test.ts)

### PR DESCRIPTION
**Pull request by ocdi**
Fixes #495 and #494 and also consolidates some utility methods used across blob, table and queues.

### CODE REVIEW

**Comment by ocdi** [comment]
I believe I have got this mostly complete on my draft PR, however the test cases need both fixing and fleshing out considerably. The `overrideHeaders` method in `testutils.ts` appears to be causing issues with my testing because it is modifying the `__proto__` and binding to a specific instance, which is no longer what I want it to be.

File: src/table/middleware/AuthenticationMiddlewareFactory.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    authenticators: IAuthenticator[]
+  ): RequestHandler {
+    return (req: Request, res: Response, next: NextFunction) => {
+      const context = new TableStorageContext(
+        res.locals,
+        DEFAULT_TABLE_CONTEXT_PATH
+      );
+      this.authenticate(req, res, authenticators)
+        .then(pass => {
+          // TODO: To support public access, we need to modify here to reject request later in handler
```
Comment (XiaoningLiu) [inline src/table/middleware/AuthenticationMiddlewareFactory.ts:23]: >// TODO: To support public access, we need to modify here to reject request later in handler [](start = 10, length = 92)

Table service doesn't support public access.

File: src/table/utils/constants.ts [hunk 2]
```diff
@@ -27,3 +33,9 @@
 
 export const RETURN_NO_CONTENT = "return-no-content";
 export const RETURN_CONTENT = "return-content";
+
+export const EMULATOR_ACCOUNT_NAME = "devstoreaccount1";
```
Comment (XiaoningLiu) [inline src/table/utils/constants.ts:37]: >EMULATOR_ACCOUNT_NAME [](start = 13, length = 21)

We should already have emulator name and key constants in src/common/... Don't need to duplicate here.

File: src/table/utils/utils.ts [hunk 3]
```diff
@@ -10,3 +13,73 @@ export function newEtag(): string {
...
+}
+
+/**
+ * Get URL query key value pairs from an URL string.
+ *
+ * @export
+ * @param {string} url
+ * @returns {{[key: string]: string}}
+ */
+export function getURLQueries(url: string): { [key: string]: string } {
```
Comment (XiaoningLiu) [inline src/table/utils/utils.ts:58]: >getURLQueries [](start = 16, length = 13)

[minor] Can you help move this into src/common/utils/utils.ts? And update blob/queue/table refer to same helper method?

File: src/table/utils/utils.ts [hunk 4]
```diff
@@ -10,3 +13,73 @@ export function newEtag(): string {
...
+/**
+ * Rounds a date off to seconds.
+ *
+ * @export
+ * @param {Date} date
+ * @param {boolean} [withMilliseconds=true] If true, YYYY-MM-DDThh:mm:ss.fffffffZ will be returned;
+ *                                          If false, YYYY-MM-DDThh:mm:ssZ will be returned.
+ * @returns {string} Date string in ISO8061 format, with or without 7 milliseconds component
+ */
+export function truncatedISO8061Date(
```
Comment (XiaoningLiu) [inline src/table/utils/utils.ts:39]: >truncatedISO8061Date [](start = 16, length = 20)

[minor] Can you help move this into src/common/utils/utils.ts? And update blob/queue/table refer to same helper method?

File: src/table/utils/utils.ts [hunk 5]
```diff
@@ -10,3 +13,73 @@ export function newEtag(): string {
...
 }
+
+/**
+ * Generates a hash signature for an HTTP request or for a SAS.
+ *
+ * @param {string} stringToSign
+ * @param {key} key
+ * @returns {string}
+ */
+export function computeHMACSHA256(stringToSign: string, key: Buffer): string {
```
Comment (XiaoningLiu) [inline src/table/utils/utils.ts:24]: >computeHMACSHA256 [](start = 16, length = 17)

[minor] Can you help move this into src/common/utils/utils.ts? And update blob/queue/table refer to same helper method?

File: src/table/errors/StorageErrorFactory.ts [hunk 6]
```diff
@@ -31,6 +31,67 @@ export default class StorageErrorFactory {
       context
     );
   }
+  public static getInvalidOperation(
```
Comment (XiaoningLiu) [inline src/table/errors/StorageErrorFactory.ts:34]: >   [](start = 0, length = 2)

[Minor] Needs an addition space here.

File: src/table/authentication/AccountSASAuthenticator.ts [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+      `AccountSASAuthenticator:validate() Start validation against account Shared Access Signature pattern.`,
+      context.contextID
+    );
+
+    this.logger.debug(
+      "AccountSASAuthenticator:validate() Getting account properties...",
+      context.contextID
+    );
+
+    // Doesn't create a BlobContext here because wants to move this class into common
```
Comment (XiaoningLiu) [inline src/table/authentication/AccountSASAuthenticator.ts:34]: >BlobContext [](start = 24, length = 11)

Check typos related to blobxxx..

File: src/table/authentication/AccountSASAuthenticator.ts [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+    const account: string = context.context.account;
+    const containerName: string | undefined = context.context.container;
```
Comment (XiaoningLiu) [inline src/table/authentication/AccountSASAuthenticator.ts:36]: >containerName [](start = 10, length = 13)

Should update to table concepts..

**Review by XiaoningLiu — COMMENTED** [review]
Thanks @ocdi! Appreicated for your contribution!

Added some comments, Seems the PR is not finalized yet, will take another look when you finish.

File: src/table/utils/utils.ts (hunk 3 above)
Comment (ocdi) [reply src/table/utils/utils.ts:58]: I did notice a fair bit of code duplication, but very happy to refactor and consolidate some of these identical methods per your suggestion here

File: src/table/middleware/AuthenticationMiddlewareFactory.ts (hunk 1 above)
Comment (ocdi) [reply src/table/middleware/AuthenticationMiddlewareFactory.ts:23]: A lot of my changes were copy/paste :-)

**Comment by ocdi** [comment]
I've addresses most of the PR comments. Added a bunch of SAS tests and solved the `__proto__` trashing issue. Need to confirm all the SAS table and account permissions are aligned correctly and fully, which is the most tedious task remaining.

**Comment by XiaoningLiu** [comment]
Thanks @ocdi, looks great! There are some conflicts, can you have a check? after that we can merge the PR.

**Review by XiaoningLiu — APPROVED** [review]

**Comment by ocdi** [comment]
I rebased, the conflicts were my own work.

**Pull request merged**